File: resources/androidstudio/app/src/main/java/com/nativephp/mobile/queue/BackgroundQueueWorker.kt
package com.nativephp.mobile.queue

import android.content.Context
import android.util.Log
import androidx.work.BackoffPolicy
import androidx.work.Constraints
import androidx.work.CoroutineWorker
import androidx.work.ExistingPeriodicWorkPolicy
import androidx.work.NetworkType
import androidx.work.PeriodicWorkRequestBuilder
import androidx.work.WorkManager
import androidx.work.WorkerParameters
import com.nativephp.mobile.bridge.PHPBridge
import com.nativephp.mobile.network.PHPRequest
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.delay
import kotlinx.coroutines.withContext
import org.json.JSONObject
import java.util.concurrent.TimeUnit

/**
 * WorkManager-based background worker for processing Laravel queue jobs.
 *
 * This worker handles job processing when the app is in the background,
 * complementing the foreground NativeQueueCoordinator. Jobs are processed
 * in batches during system-allocated background work windows.
 *
 * The worker integrates with the existing queue system:
 * - Uses the same `/_native/queue/work` endpoint as foreground processing
 * - Respects the same throttling configuration from PHP
 * - Re-schedules itself when jobs remain after processing
 */
class BackgroundQueueWorker(
    context: Context,
    params: WorkerParameters
) : CoroutineWorker(context, params) {

    companion object {
        private const val TAG = "BackgroundQueueWorker"
        private const val WORK_NAME = "nativephp_queue_processing"

        // Default configuration
        private const val DEFAULT_INTERVAL_MINUTES = 15L
        private const val MAX_PROCESSING_TIME_MS = 9 * 60 * 1000L  // 9 minutes (leave buffer)
        private const val MAX_JOBS_PER_SESSION = 50
        private const val DEFAULT_MIN_DELAY_MS = 100L

        /**
         * Schedule periodic background queue processing.
         * Called when the app enters background.
         *
         * @param context Application context
         * @param intervalMinutes Minimum interval between work executions (default 15)
         */
        fun schedule(context: Context, intervalMinutes: Long = DEFAULT_INTERVAL_MINUTES) {
            val constraints = Constraints.Builder()
                .setRequiredNetworkType(NetworkType.NOT_REQUIRED)  // SQLite is local
                .build()

            val workRequest = PeriodicWorkRequestBuilder<BackgroundQueueWorker>(
                intervalMinutes, TimeUnit.MINUTES
            )
                .setConstraints(constraints)
                .setBackoffCriteria(
                    BackoffPolicy.EXPONENTIAL,
                    10, TimeUnit.SECONDS
                )
                .build()

            WorkManager.getInstance(context).enqueueUniquePeriodicWork(
                WORK_NAME,
                ExistingPeriodicWorkPolicy.KEEP,
                workRequest
            )

            Log.d(TAG, "Scheduled background queue processing with ${intervalMinutes}min interval")
        }

        /**
         * Cancel any pending background work.
         * Called when the app returns to foreground.
         */
        fun cancel(context: Context) {
            WorkManager.getInstance(context).cancelUniqueWork(WORK_NAME)
            Log.d(TAG, "Cancelled background queue processing")
        }
    }

    override suspend fun doWork(): Result = withContext(Dispatchers.IO) {
        Log.d(TAG, "Starting background queue processing")

        val phpBridge = PHPBridge.getInstance()
        if (phpBridge == null) {
            Log.w(TAG, "PHPBridge not available, retrying later")
            return@withContext Result.retry()
        }

        var minDelayMs = DEFAULT_MIN_DELAY_MS
        var jobsProcessed = 0
        var remainingJobs = 0
        val startTime = System.currentTimeMillis()

        try {
            // First, check queue status and get config
            val statusResponse = makeRequest(phpBridge, "/_native/queue/status", "GET")
            if (statusResponse != null) {
                val pending = statusResponse.optInt("pending", 0)
                if (pending == 0) {
                    Log.d(TAG, "Queue is empty, nothing to process")
                    return@withContext Result.success()
                }

                // Apply config from PHP
                val config = statusResponse.optJSONObject("config")
                if (config != null) {
                    minDelayMs = config.optLong("min_delay", DEFAULT_MIN_DELAY_MS)
                }
            }

            // Process jobs in a loop
            while (true) {
                // Check time limit
                val elapsed = System.currentTimeMillis() - startTime
                if (elapsed >= MAX_PROCESSING_TIME_MS) {
                    Log.d(TAG, "Time limit reached after $jobsProcessed jobs")
                    break
                }

                // Check job limit
                if (jobsProcessed >= MAX_JOBS_PER_SESSION) {
                    Log.d(TAG, "Job limit reached ($MAX_JOBS_PER_SESSION)")
                    break
                }

                // Check if stopped
                if (isStopped) {
                    Log.d(TAG, "Worker stopped after $jobsProcessed jobs")
                    break
                }

                // Process one job
                val workResponse = makeRequest(phpBridge, "/_native/queue/work", "POST")
                if (workResponse == null) {
                    Log.e(TAG, "Failed to get work response")
                    break
                }

                val processed = workResponse.optBoolean("processed", false)
                remainingJobs = workResponse.optInt("pending", 0)

                if (processed) {
                    jobsProcessed++

                    // Log job info
                    val jobData = workResponse.optJSONObject("job")
                    val jobName = jobData?.optString("name", "unknown") ?: "unknown"
                    Log.d(TAG, "Processed job $jobsProcessed: $jobName")

                    // Continue with next job after delay if more jobs remain
                    if (remainingJobs > 0) {
                        delay(minDelayMs)
                    } else {
                        Log.d(TAG, "Queue empty after $jobsProcessed jobs")
                        break
                    }
                } else {
                    // Queue was empty
                    Log.d(TAG, "Queue empty (processed=false)")
                    break
                }
            }

            Log.d(TAG, "Background processing complete: $jobsProcessed jobs processed, $remainingJobs remaining")
            Result.success()

        } catch (e: Exception) {
            Log.e(TAG, "Error during background processing", e)
            Result.retry()
        }
    }

    /**
     * Make an HTTP request to the Laravel queue endpoint.
     */
    private fun makeRequest(phpBridge: PHPBridge, path: String, method: String): JSONObject? {
        return try {
            val request = PHPRequest(
                url = path,
                method = method,
                headers = mapOf(
                    "Accept" to "application/json",
                    "Content-Type" to "application/json"
                )
            )

            val response = phpBridge.handleLaravelRequest(request)

            // Parse response - extract body from HTTP response
            val parts = response.split("\r\n\r\n", limit = 2)
            if (parts.size >= 2) {
                JSONObject(parts[1])
            } else {
                null
            }
        } catch (e: Exception) {
            Log.e(TAG, "Request failed: ${e.message}")
            null
        }
    }
}

File: resources/xcode/NativePHP/Queue/BackgroundQueueWorker.swift
import Foundation
import BackgroundTasks

/// Manages background queue processing using iOS BGTaskScheduler.
///
/// This worker handles job processing when the app is in the background,
/// complementing the foreground NativeQueueCoordinator. Jobs are processed
/// in batches during system-allocated background time windows (up to 30 seconds).
///
/// The worker integrates with the existing queue system:
/// - Uses the same `/_native/queue/work` endpoint as foreground processing
/// - Respects the same throttling configuration
/// - Re-schedules itself when jobs remain after processing
final class BackgroundQueueWorker {
    static let shared = BackgroundQueueWorker()

    /// Task identifier for BGTaskScheduler - must match Info.plist
    private var taskIdentifier: String {
        let bundleId = Bundle.main.bundleIdentifier ?? "com.nativephp.app"
        return "\(bundleId).queue-processing"
    }

    // MARK: - Configuration

    /// Maximum time to spend processing jobs (leave buffer for cleanup)
    private let maxProcessingTime: TimeInterval = 25.0

    /// Maximum jobs to process per background session
    private let maxJobsPerSession: Int = 50

    /// Minimum delay between job processing
    private var minDelayBetweenJobs: TimeInterval = 0.1

    // MARK: - State

    private var isProcessing = false
    private var shouldStop = false
    private var pendingJobCount = 0

    private init() {}

    // MARK: - Public API

    /// Register the background task handler.
    /// Must be called during app launch, before applicationDidFinishLaunching returns.
    func register() {
        BGTaskScheduler.shared.register(
            forTaskWithIdentifier: taskIdentifier,
            using: nil
        ) { [weak self] task in
            self?.handleBackgroundTask(task as! BGProcessingTask)
        }

        DebugLogger.shared.log("BackgroundQueue: Registered task handler for \(taskIdentifier)")
    }

    /// Schedule a background task if there are pending jobs.
    /// Called when the app enters background.
    func scheduleIfNeeded() {
        // First check if we have pending jobs via the queue status
        checkPendingJobsAndSchedule()
    }

    /// Cancel any pending background task.
    /// Called when the app returns to foreground.
    func cancelPendingTask() {
        BGTaskScheduler.shared.cancel(taskRequestWithIdentifier: taskIdentifier)
        DebugLogger.shared.log("BackgroundQueue: Cancelled pending background task")
    }

    // MARK: - Private

    private func checkPendingJobsAndSchedule() {
        // Make a quick status check to see if there are jobs to process
        PHPSchemeHandler.phpSerialQueue.async { [weak self] in
            guard let self = self else { return }

            let requestData = RequestData(
                method: "GET",
                uri: "/_native/queue/status",
                data: nil,
                query: nil,
                headers: ["Accept": "application/json"]
            )

            guard let response = NativePHPApp.laravel(request: requestData) else {
                return
            }

            // Parse response
            let components = response.components(separatedBy: "\r\n\r\n")
            guard components.count >= 2,
                  let data = components[1].data(using: .utf8),
                  let json = try? JSONSerialization.jsonObject(with: data) as? [String: Any],
                  let pending = json["pending"] as? Int else {
                return
            }

            self.pendingJobCount = pending

            // Update config if available
            if let config = json["config"] as? [String: Any],
               let minDelay = config["min_delay"] as? Int {
                self.minDelayBetweenJobs = TimeInterval(minDelay) / 1000.0
            }

            // Only schedule if there are jobs waiting
            if pending > 0 {
                self.scheduleBackgroundTask()
            } else {
                DebugLogger.shared.log("BackgroundQueue: No pending jobs, skipping background task schedule")
            }
        }
    }

    private func scheduleBackgroundTask() {
        let request = BGProcessingTaskRequest(identifier: taskIdentifier)
        request.requiresNetworkConnectivity = false  // SQLite is local
        request.requiresExternalPower = false

        do {
            try BGTaskScheduler.shared.submit(request)
            DebugLogger.shared.log("BackgroundQueue: Scheduled background task with \(pendingJobCount) pending jobs")
        } catch {
            DebugLogger.shared.log("BackgroundQueue: Failed to schedule task: \(error.localizedDescription)")
        }
    }

    private func handleBackgroundTask(_ task: BGProcessingTask) {
        DebugLogger.shared.log("BackgroundQueue: Starting background processing")

        shouldStop = false
        isProcessing = true

        // Set expiration handler to gracefully stop
        task.expirationHandler = { [weak self] in
            DebugLogger.shared.log("BackgroundQueue: System requested expiration")
            self?.shouldStop = true
        }

        // Process jobs in the background
        processBackgroundBatch { [weak self] success, jobsProcessed, remainingJobs in
            guard let self = self else { return }

            self.isProcessing = false

            DebugLogger.shared.log("BackgroundQueue: Completed - processed \(jobsProcessed) jobs, \(remainingJobs) remaining")

            // Re-schedule if jobs remain
            if remainingJobs > 0 {
                self.pendingJobCount = remainingJobs
                self.scheduleBackgroundTask()
            }

            task.setTaskCompleted(success: success)
        }
    }

    private func processBackgroundBatch(completion: @escaping (Bool, Int, Int) -> Void) {
        let startTime = Date()
        var jobsProcessed = 0
        var remainingJobs = 0
        var hadError = false

        func processNextJob() {
            // Check stop conditions
            guard !shouldStop else {
                completion(!hadError, jobsProcessed, remainingJobs)
                return
            }

            // Check time limit
            let elapsed = Date().timeIntervalSince(startTime)
            guard elapsed < maxProcessingTime else {
                DebugLogger.shared.log("BackgroundQueue: Time limit reached after \(jobsProcessed) jobs")
                completion(!hadError, jobsProcessed, remainingJobs)
                return
            }

            // Check job limit
            guard jobsProcessed < maxJobsPerSession else {
                DebugLogger.shared.log("BackgroundQueue: Job limit reached (\(maxJobsPerSession))")
                completion(!hadError, jobsProcessed, remainingJobs)
                return
            }

            // Process one job
            PHPSchemeHandler.phpSerialQueue.async { [weak self] in
                guard let self = self else {
                    completion(false, jobsProcessed, remainingJobs)
                    return
                }

                let requestData = RequestData(
                    method: "POST",
                    uri: "/_native/queue/work",
                    data: nil,
                    query: nil,
                    headers: [
                        "Accept": "application/json",
                        "Content-Type": "application/json"
                    ]
                )

                guard let response = NativePHPApp.laravel(request: requestData) else {
                    hadError = true
                    completion(false, jobsProcessed, remainingJobs)
                    return
                }

                // Parse response
                let components = response.components(separatedBy: "\r\n\r\n")
                guard components.count >= 2,
                      let data = components[1].data(using: .utf8),
                      let json = try? JSONSerialization.jsonObject(with: data) as? [String: Any] else {
                    hadError = true
                    completion(false, jobsProcessed, remainingJobs)
                    return
                }

                let processed = json["processed"] as? Bool ?? false
                remainingJobs = json["pending"] as? Int ?? 0

                if processed {
                    jobsProcessed += 1

                    // Extract job info for logging
                    if let jobData = json["job"] as? [String: Any],
                       let jobName = jobData["name"] as? String {
                        DebugLogger.shared.log("BackgroundQueue: Processed job \(jobsProcessed): \(jobName)")
                    }

                    // Continue with next job after delay
                    if remainingJobs > 0 && !self.shouldStop {
                        DispatchQueue.global(qos: .utility).asyncAfter(deadline: .now() + self.minDelayBetweenJobs) {
                            processNextJob()
                        }
                    } else {
                        // Queue empty or stopped
                        completion(!hadError, jobsProcessed, remainingJobs)
                    }
                } else {
                    // Queue was empty
                    DebugLogger.shared.log("BackgroundQueue: Queue empty after \(jobsProcessed) jobs")
                    completion(!hadError, jobsProcessed, 0)
                }
            }
        }

        // Start processing
        processNextJob()
    }
}
